File: src/main/java/backend/academy/image/drawer/MultiThreadDrawer.java
package backend.academy.image.drawer;

import backend.academy.pixel.Pixel;
import backend.academy.pixel.Pixels;
import java.awt.Color;
import java.awt.image.BufferedImage;
import java.util.concurrent.ForkJoinPool;
import java.util.concurrent.RecursiveTask;

public class MultiThreadDrawer implements Drawer {

    @Override
    public BufferedImage draw(ImageSettings imageSettings, Pixels pixels) {
        // Определяем размеры изображения
        int width = imageSettings.width();
        int height = imageSettings.height();

        // Создаем пустое изображение
        BufferedImage image = new BufferedImage(width, height, BufferedImage.TYPE_INT_RGB);

        // Создаем ForkJoinPool и запускаем главную задачу
        try (var pool = ForkJoinPool.commonPool()) {
            pool.invoke(new DrawTask(0, pixels.pixels().length, image, pixels, width, height));
        }
        return image;
    }

    /**
     * Рекурсивная задача для отрисовки
     */
    private static class DrawTask extends RecursiveTask<Void> {
        private static final int THRESHOLD = 50; // Максимальное количество строк для одной задачи

        private final int startRow; // Начальная строка
        private final int endRow;   // Конечная строка (не включительно)
        private final BufferedImage image;
        private final Pixels pixels;
        private final int width;
        private final int height;

        DrawTask(int startRow, int endRow, BufferedImage image, Pixels pixels, int width, int height) {
            this.startRow = startRow;
            this.endRow = endRow;
            this.image = image;
            this.pixels = pixels;
            this.width = width;
            this.height = height;
        }

        @Override
        protected Void compute() {
            // Если количество строк меньше порога, выполняем задачу напрямую
            if (endRow - startRow <= THRESHOLD) {
                drawPixels();
                return null;
            }

            // Делим задачу на две подзадачи
            int mid = (startRow + endRow) / 2;
            DrawTask topTask = new DrawTask(startRow, mid, image, pixels, width, height);
            DrawTask bottomTask = new DrawTask(mid, endRow, image, pixels, width, height);

            // Запускаем подзадачи параллельно и дожидаемся их завершения
            invokeAll(topTask, bottomTask);
            return null;
        }

        // Метод для непосредственной отрисовки пикселей
        private void drawPixels() {
            Pixel[][] pixelArray = pixels.pixels();

            for (int row = startRow; row < endRow; row++) {
                for (int col = 0; col < pixelArray[row].length; col++) {
                    Pixel pixel = pixelArray[row][col];

                    Color color = new Color(pixel.color().red(), pixel.color().green(), pixel.color().blue());

                    // Преобразуем координаты пикселя в пространство изображения
                    double normalizedX = (double) col / pixelArray[row].length; // Нормализация в диапазон [0, 1]
                    double normalizedY = (double) row / pixelArray.length;      // Нормализация в диапазон [0, 1]

                    int x = (int) Math.round(normalizedX * (width - 1)); // Масштабируем в [0, width - 1]
                    int y = (int) Math.round(normalizedY * (height - 1)); // Масштабируем в [0, height - 1]

                    // Устанавливаем цвет пикселя в изображении
                    if (x >= 0 && x < width && y >= 0 && y < height) {
                        image.setRGB(x, y, color.getRGB());

                    }
                }
            }
        }
    }
}

File: src/main/kotlin/backend/academy/image/drawer/MultiThreadDrawer.kt
package backend.academy.backend.academy.image.drawer

import backend.academy.image.drawer.Drawer
import backend.academy.image.drawer.ImageSettings
import backend.academy.pixel.Pixels
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import java.awt.Color
import java.awt.image.BufferedImage
import java.util.concurrent.ForkJoinPool
import java.util.concurrent.RecursiveTask

class MultiThreadDrawer : Drawer {
    override fun draw(imageSettings: ImageSettings, pixels: Pixels): BufferedImage {
        // Определяем размеры изображения
        val width = imageSettings.width
        val height = imageSettings.height

        // Создаем пустое изображение
        val image = BufferedImage(width, height, BufferedImage.TYPE_INT_RGB)
        val task = DrawTask(image, pixels,width,height)
        runBlocking {
            task.compute()
        }

        return image
    }

    /**
     * Рекурсивная задача для отрисовки
     */
    private class DrawTask(
        private val image: BufferedImage,
        private val pixels: Pixels,
        private val width: Int,
        private val height: Int,
        private val startRow: Int = 0,
        private val endRow: Int = pixels.getPixels().size // Конечная строка (не включительно)
    ) {
        suspend fun compute() {
            // Если количество строк меньше порога, выполняем задачу напрямую
            if (endRow - startRow <= THRESHOLD) {
                drawPixels()
            }else {
                // Делим задачу на две подзадачи
                val mid = (startRow + endRow) / 2
                val topTask = DrawTask( image, pixels, width, height,startRow, mid)
                val bottomTask = DrawTask( image, pixels, width, height,mid, endRow)
                coroutineScope {
                    val job1 = launch {
                        topTask.compute()
                    }
                    val job2 = launch {
                        bottomTask.compute()
                    }
                    job1.join()
                    job2.join()
                }
            }
        }

        // Метод для непосредственной отрисовки пикселей
        fun drawPixels() {
            val pixelArray = pixels.getPixels()

            for (row in startRow..<endRow) {
                for (col in pixelArray[row].indices) {
                    val pixel = pixelArray[row][col]

                    val color = Color(pixel.color.red, pixel.color.green, pixel.color.blue)

                    // Преобразуем координаты пикселя в пространство изображения
                    val normalizedX = col.toDouble() / pixelArray[row].size // Нормализация в диапазон [0, 1]
                    val normalizedY = row.toDouble() / pixelArray.size // Нормализация в диапазон [0, 1]

                    val x = Math.round(normalizedX * (width - 1)).toInt() // Масштабируем в [0, width - 1]
                    val y = Math.round(normalizedY * (height - 1)).toInt() // Масштабируем в [0, height - 1]

                    // Устанавливаем цвет пикселя в изображении
                    if (x >= 0 && x < width && y >= 0 && y < height) {
                        image.setRGB(x, y, color.rgb)
                    }
                }
            }
        }

        companion object {
            private const val THRESHOLD = 50 // Максимальное количество строк для одной задачи
        }
    }
}